### PR TITLE
Add JAX Hessian gotcha to `enable_return` docs

### PR DESCRIPTION
**Pull request by antalszava**
* Creates a new `JAX Usage Details` section;
* Moves the existing JAX examples to the new section;
* Renames existing `Usage Details for <ML framework>` section titles to `<ML framework> Usage Details`;
* Adds the gotchas on JAX Hessian.

### CODE REVIEW

File: pennylane/return_types.py [hunk 1]
```diff
@@ -181,104 +181,8 @@ def circuit(a, b):
 
             AttributeError: 'tuple' object has no attribute 'shape'
 
-        **JAX interface upgrades: higher-order derivatives and mixing measurements**
```
Comment (antalszava) [inline pennylane/return_types.py:184]: This part was simply pasted in the new JAX Usage Details section.

File: pennylane/return_types.py [hunk 2]
```diff
@@ -181,104 +181,8 @@ def circuit(a, b):
...
-         (Array([[ 0.07677898,  0.0563341 ],
-                       [ 0.07238522, -1.830669  ]], dtype=float32),
-          Array([[[-4.9707499e-01,  2.9999996e-04],
-                        [-6.2500127e-04,  1.2500001e-04]],
-                       [[ 4.9707499e-01, -2.9999996e-04],
-                        [ 6.2500127e-04, -1.2500001e-04]]], dtype=float32)))
-
     .. details::
-        :title: Usage Details for Autograd
+        :title: Autograd Usage Details
```
Comment (antalszava) [inline pennylane/return_types.py:185]: Personally, I think having the titles be different in the first word can boost looking and browsing through the page as a whole.

File: pennylane/return_types.py [hunk 3]
```diff
@@ -588,6 +492,147 @@ def circuit_stack(a, b):
...
+                      [[ 0.5,  0. ],
+                       [ 0. ,  0. ]]], dtype=float32)),
+         (Array([[ 0.07677898,  0.0563341 ],
+                       [ 0.07238522, -1.830669  ]], dtype=float32),
+          Array([[[-4.9707499e-01,  2.9999996e-04],
+                        [-6.2500127e-04,  1.2500001e-04]],
+                       [[ 4.9707499e-01, -2.9999996e-04],
+                        [ 6.2500127e-04, -1.2500001e-04]]], dtype=float32)))
+
+    Note that failing to set the ``max-diff`` with jitting will raise a somewhat unrelated error:
```
Comment (antalszava) [inline pennylane/return_types.py:592]: From this point on, this is a new part of the section.

Comment (eddddddy) [reply pennylane/return_types.py:592]: Is this meant to be in the JAX details section still? If so, I think it needs an extra indent to render correctly.

File: pennylane/return_types.py [hunk 4]
```diff
@@ -588,6 +492,147 @@ def circuit_stack(a, b):
...
+              return qml.expval(qml.PauliZ(wires=0)), qml.probs(wires=[0, 1])
+
+            x = jax.numpy.array([0.1, 0.2, 0.3])
+
+        >>> jax.jacobian(circuit)(x)
+        (Array([-9.9833414e-02, -7.4505806e-09,  6.9285655e-10], dtype=float32),
+         Array([[-4.9419206e-02, -9.9086545e-02,  3.4938008e-09],
+                      [-4.9750542e-04,  9.9086538e-02,  1.2768372e-10],
+                      [ 4.9750548e-04,  2.4812977e-04,  4.8371929e-13],
+                      [ 4.9419202e-02, -2.4812980e-04,  2.6696912e-11]],            dtype=float32))
```
Comment (eddddddy) [inline pennylane/return_types.py:550]:
```suggestion
         Array([[-4.9419206e-02, -9.9086545e-02,  3.4938008e-09],
                [-4.9750542e-04,  9.9086538e-02,  1.2768372e-10],
                [ 4.9750548e-04,  2.4812977e-04,  4.8371929e-13],
                [ 4.9419202e-02, -2.4812980e-04,  2.6696912e-11]],            dtype=float32))
```

File: pennylane/return_types.py [hunk 5]
```diff
@@ -588,6 +492,147 @@ def circuit_stack(a, b):
...
+          Array([[[-0.5,  0. ],
+                       [ 0. ,  0. ]],
+                      [[ 0.5,  0. ],
+                       [ 0. ,  0. ]]], dtype=float32)),
+         (Array([[ 0.07677898,  0.0563341 ],
+                       [ 0.07238522, -1.830669  ]], dtype=float32),
+          Array([[[-4.9707499e-01,  2.9999996e-04],
+                        [-6.2500127e-04,  1.2500001e-04]],
+                       [[ 4.9707499e-01, -2.9999996e-04],
+                        [ 6.2500127e-04, -1.2500001e-04]]], dtype=float32)))
```
Comment (eddddddy) [inline pennylane/return_types.py:590]: Some formatting issues here.

File: pennylane/return_types.py [hunk 6]
```diff
@@ -588,6 +492,147 @@ def circuit_stack(a, b):
...
+
+        >>> jax.hessian(circuit)(params)
+        ~/anaconda3/lib/python3.8/site-packages/jax/_src/callback.py in pure_callback_jvp_rule(***failed resolving arguments***)
+             53 def pure_callback_jvp_rule(*args, **kwargs):
+             54   del args, kwargs
+        ---> 55   raise ValueError(
+             56       "Pure callbacks do not support JVP. "
+             57       "Please use `jax.custom_jvp` to use callbacks while taking gradients.")
+
+        ValueError: Pure callbacks do not support JVP. Please use `jax.custom_jvp` to use callbacks while taking gradients.
```
Comment (eddddddy) [inline pennylane/return_types.py:622]: It seems Sphinx doesn't like newlines in tracebacks, and renders this as outside of the console output. I think we just have to accept this limitation and remove the newline here.

File: pennylane/return_types.py [hunk 7]
```diff
@@ -588,6 +492,147 @@ def circuit_stack(a, b):
...
+
+    Correctly specifying ``max_diff=2`` as a QNode argument helps compute the Hessian:
+
+        .. code-block:: python
```
Comment (eddddddy) [inline pennylane/return_types.py:626]: Unused `code-block` directive.

File: pennylane/return_types.py [hunk 8]
```diff
@@ -588,6 +492,147 @@ def circuit_stack(a, b):
...
+
+        >>> jax.hessian(circuit)(params)
+        (Array([[ 0.06617428,  0.07165382],
+                [ 0.07165382, -1.8348092 ]], dtype=float64),
+         Array([[[-4.974e-01, -2.025e-03],
+                 [-2.025e-03,  1.000e-04]],
+
+                [[ 4.974e-01,  2.025e-03],
+                 [ 2.025e-03, -1.000e-04]]], dtype=float64))
```
Comment (eddddddy) [inline pennylane/return_types.py:635]: Same issue with newlines.

**Review by rmoyard — APPROVED** [review]
:100: Thanks Antal!

**Review by eddddddy — APPROVED** [review]
Thanks 💯

**Pull request merged**